serverless/email/status/main.go

```diff
@@ -0,0 +1,150 @@
+}
+
+type bounce struct {
+	Timestamp time.Time `json:"timestamp"`
+}
+
+type complaint struct {
+	Timestamp time.Time `json:"timestamp"`
+}
+
+type delivery struct {
+	Timestamp time.Time `json:"timestamp"`
+}
+
+type deliveryDelay struct {
+	Timestamp time.Time `json:"timestamp"`
+}
+
+func handler(ctx context.Context, snsEvent events.SNSEvent) {
+	// Process the SES delivery notifications included in the SNS event.
+	// An event may include notifications one or multiple recipients
+	// A single recipient might receive multiple notifications
+	// https://docs.aws.amazon.com/ses/latest/dg/notification-contents.html
+	for _, record := range snsEvent.Records {
+		var notification sesNotification
+		err := json.Unmarshal([]byte(record.SNS.Message), &notification)
+		if err != nil {
+			logger.Error().Err(err).Msg("failed to unmarshal notification")
+			continue
+		}
+
+		// Check the status type and get the associated timestamp
+		// Bounce, Complaint, Delivery, Send, Reject, Open, Click, Rendering Failure, DeliveryDelay, or Subscription.
+		var statusTimestamp time.Time
+		switch notification.EventType {
+		case "Bounce":
+			statusTimestamp = notification.Bounce.Timestamp
+		case "Complaint":
+			statusTimestamp = notification.Complaint.Timestamp
+		case "Delivery":
+			statusTimestamp = notification.Delivery.Timestamp
+		case "Send":
+		case "Reject":
+		case "Open":
+			continue // Skip "Open" events
+		case "Click":
+			continue // Skip "Click" events
+		case "Rendering Failure":
+		case "DeliveryDelay":
+			statusTimestamp = notification.DeliveryDelay.Timestamp
+		case "Subscription":
+		default:
+			logger.Warn().Msg("unknown event type " + notification.EventType)
+		}
+
+		// Get Idempotency key from tags to use as partition key, skip if it is not present
+		var idempotencyKey string
+		for _, tag := range notification.Mail.Tags {
+			if *(tag.Name) == "idempotencyKey" {
+				idempotencyKey = *(tag.Value)
+				break
+			}
+		}
+		if idempotencyKey == "" {
+			logger.Warn().Msg("missing idempotency ID from email " + notification.Mail.MessageID)
+			continue
+		}
+
+		// Write status to database
+		item := map[string]types.AttributeValue{
+			"FtxIdempotencyKey": &types.AttributeValueMemberS{Value: idempotencyKey},
+			"SesMessageId":      &types.AttributeValueMemberS{Value: notification.Mail.MessageID},
+			"StatusId":          &types.AttributeValueMemberS{Value: uuid.NewString()},
+			"StatusType":        &types.AttributeValueMemberS{Value: notification.EventType},
+			"CreatedAt":         &types.AttributeValueMemberN{Value: strconv.FormatInt(time.Now().UTC().Unix(), 10)},
+		}
+		// Include StatusTs only if it has a value
+		if !statusTimestamp.IsZero() {
+			item["StatusTs"] = &types.AttributeValueMemberN{Value: strconv.FormatInt(statusTimestamp.Unix(), 10)}
+		}
+		_, err = dynamoClient.PutItem(ctx, &dynamodb.PutItemInput{
+			TableName: statusTable,
+			Item:      item,
+		})
+		if err != nil {
+			logger.Error().Err(err).Msg(
```

Comment:
There is a format on the logger which can be quite handy,  `logger.Error().Err(err).Msgf("failed to write status to dynamodb for messageID %s", notification.Mail.MessageID)`

Reply:
https://github.com/brave-intl/bat-go/pull/1640